### PR TITLE
Extend realloc-in-place to arrays of non-POD data

### PR DESCRIPTION
**Pull request by bradcray**
* implement an @mppf observation that we _can_ realloc non-POD data,
  we simply need to call deinit on the elements that are going away before
  invoking realloc; this increases the number of cases that can use realloc

* add a config param to support disabling the realloc code path since
  it turned out not to be quite as simple as originally hoped (in case
  we run into users who have trouble with it in the field).

* rename allocD to reallocD in the array resize code, because I
  could never remember whether "alloc" meant "the domain describing
  the current allocation or the new allocation".  Hopefully the
  new term makes it clearer that it's the latter (?).

* add a comment about a potential opportunity to parallelize array
  element deinitialization in the future (related to issue #15215).

### CODE REVIEW

**Comment by bradcray** [comment]
@mppf: I'm curious whether you would review this as the person who (sagely) pointed out that I closed the door on non-POD types too quickly in PR #15202 last night?

Also tagging @gbtitus and @ronawho due to their involvement in #15177

**Comment by bradcray** [comment]
@mppf: Do you think you could find time to review this today?

File: modules/internal/DefaultRectangular.chpl [hunk 1]
```diff
@@ -1340,33 +1343,32 @@ module DefaultRectangular {
...
-            dom.dsiNumIndices > 0 && allocD.size > 0 &&
-            (isPODType(eltType) || allocD.size > dom.dsiNumIndices)) {
+        if (!disableArrRealloc && rank == 1 &&
+            reallocD.low == dom.dsiLow && reallocD.stride == dom.dsiStride &&
+            dom.dsiNumIndices > 0 && reallocD.size > 0) {
+
           if reportInPlaceRealloc then
             writeln("reallocating in-place");
 
-          sizesPerDim(1) = allocD.dsiDim(1).size;
```
Comment (mppf) [inline modules/internal/DefaultRectangular.chpl]: Why is this line deleted? I don't see anything about it in the PR message.

Comment (bradcray) [reply modules/internal/DefaultRectangular.chpl]: Huh, I suspect that was a mistake when I was renaming allocD to reallocD.  Good catch.
